cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java

```diff
@@ -0,0 +1,119 @@
+        String desResourceGroupName;
+        AzureClient azureClient = azureClientService.getClient(diskEncryptionSetCreationRequest.getCloudCredential());
+
+        Matcher matcher = ENCRYPTION_KEY_URL_PATTERN.matcher(diskEncryptionSetCreationRequest.getEncryptionKeyUrl());
+        if (matcher.matches()) {
+            vaultName = matcher.group(1);
+        } else {
+            throw new IllegalArgumentException("vaultName cannot be fetched from encryptionKeyUrl. encryptionKeyUrl should be of format - " +
+                    "'https://<vaultName>.vault.azure.net/keys/<keyName>/<keyVersion>'");
+        }
+        if (diskEncryptionSetCreationRequest.isSingleResourceGroup()) {
+            desResourceGroupName = diskEncryptionSetCreationRequest.getResourceGroup();
+            vaultResourceGroupName = desResourceGroupName;
+        } else {
+            throw new IllegalArgumentException("Customer Managed Key Encryption for managed Azure disks is supported for single resource group only.");
+        }
+        String sourceVaultId = String.format("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.KeyVault/vaults/%s",
+                azureClient.getCurrentSubscription().subscriptionId(), vaultResourceGroupName, vaultName);
+
+        CreatedDiskEncryptionSet diskEncryptionSet = getOrCreateDiskEncryptionSetOnCloud(
+                azureClient,
+                desResourceGroupName,
+                sourceVaultId,
+                diskEncryptionSetCreationRequest);
+        azureClient.grantKeyVaultAccessPolicyToServicePrincipal(vaultResourceGroupName, vaultName,
+                diskEncryptionSet.getDiskEncryptionSetPrincipalId());
+        return diskEncryptionSet;
+    }
+
+    private CreatedDiskEncryptionSet getOrCreateDiskEncryptionSetOnCloud(AzureClient azureClient, String desResourceGroupName,
+            String sourceVaultId, DiskEncryptionSetCreationRequest diskEncryptionSetCreationRequest) {
+        String diskEncryptionSetName = azureUtils.generateDesNameByNameAndId(
+                String.format("%s-DES-", diskEncryptionSetCreationRequest.getEnvironmentName()),
+                diskEncryptionSetCreationRequest.getId());
+        LOGGER.info("Checking if Disk Encryption Set {} exists on cloud Azure", diskEncryptionSetName);
+        DiskEncryptionSetInner createdSet = azureClient.getDiskEncryptionSet(desResourceGroupName, diskEncryptionSetName);
+        if (createdSet == null) {
+            LOGGER.info("Creating Disk Encryption Set {}", diskEncryptionSetName);
+            createdSet = azureClient.createDiskEncryptionSet(diskEncryptionSetName, diskEncryptionSetCreationRequest.getEncryptionKeyUrl(),
+                    diskEncryptionSetCreationRequest.getRegion().getRegionName(), desResourceGroupName, sourceVaultId,
+                    diskEncryptionSetCreationRequest.getTags());
+        }
+        if (createdSet != null) {
+            CloudResource desCloudResource = CloudResource.builder()
```

Comment:
log for diskencryption set "name" exist on azure side. btw do we need to check not just the name but the resource group ?

Reply:
`azureClient.getDiskEncryptionSet` checks if the des exists with that name in provided resourceGroup.